Start the lang key sorting (#7643)
GTNewHorizons/GT5-Unofficial: docs/Localization.md

# Localization Guide

This document is the canonical reference for all localization work in GT5-Unofficial.
If you are touching the lang file, adding translatable strings in Java, or doing a localization
rework, read this first.

---

## 1. Java API - Which Method to Use

### 1.1 `StatCollector.translateToLocal(key)`
Use for any text that is **rendered on the client side**: item tooltips, item names, GUI labels,

> WAILA/NEI output, anything inside a `@SideOnly(Side.CLIENT)` context.

Reviewer: Also, it's fine to use it for only logging on the server (but it shouldn't be added for the server only, since the server only ever uses en_US.)

Author: What do you think about adding a note like:

> **Server-only logging:** It's fine to reuse `StatCollector.translateToLocal` for text that is
> only ever written to the server log, since that's a convenient source of an already-translated
> string. However, do not introduce a key or call `StatCollector` *for the sole purpose of*
> server-only logging - the server only ever loads `en_US.lang`, so there is no localization
> benefit to justify it.

Reviewer: Try this:
> **Server-only logging:** It's fine to call `StatCollector.translateToLocal` on the server for text 
> that is only ever written to the server log, since sometimes you need to log user-visible text.
> However, do not introduce translation keys *for the sole purpose of* server-only logging - 
> dedicated servers only ever load `en_US.lang`, so there is no localization benefit to justify it.
> (Do note that server threads on LAN & Singleplayer will use the host client's language instead.)

```java
import net.minecraft.util.StatCollector;

String name = StatCollector.translateToLocal("gt.item.item_name.name");
```

### 1.2 `StatCollector.translateToLocalFormatted(key, args...)`
Same as above but with format substitution. Use when the translated string contains `%s` or `%d`

> placeholders.

Reviewer: Note: this is different from just using `String.format` plus `StatCollector.translateToLocal`. `String.format` will throw an error if the number and type of parameters doesn't match the placeholders, and `translateToLocalFormatted` will catch that error and return "Format error" instead of crashing your program.

> 

Reviewer: ```suggestion

Avoid using `String.format` with translated strings whenever possible. `String.format` can throw
`IllegalFormatException` when given the wrong number of parameters, which could happen due
to a translator's mistake. `translateToLocalFormatted` will catch this exception and return
`Format error: [unformatted text]` instead.

```

```java
String desc = StatCollector.translateToLocalFormatted("gt.item.item_name.tooltip", 16000);
// lang: gt.item.item_name.tooltip=Capacity: %d L
```

**Never use these two methods in code that runs server-side for chat output.** On a dedicated

> server the lang file may not be loaded, so the key will be returned as-is instead of translated.

Reviewer: This is wrong. A dedicated server always loads the en_US lang file.

Author: removed

Reviewer: Nope, not as far as I can tell. Maybe you forgot to push?

> server the lang file may not be loaded, so the key will be returned as-is instead of translated.

Reviewer: ```suggestion
**Never use these two methods in code that runs server-side for chat output.** On a server,
the loaded language may differ from the client's, so the message will be translated to 
the wrong language. Use `GTUtility.sendChatTrans` or `ChatComponentTranslation` instead.
```

### 1.3 `GTUtility.sendChatTrans(player, key, args...)` - Chat Messages
Use for sending a single translated message to a player. Internally wraps `ChatComponentTranslation`,
which ships the raw key over the network and lets the **client** translate it. This works correctly
in both singleplayer and multiplayer.

```java
import gregtech.api.util.GTUtility;

// Server-side (inside a TileEntity, Command, Event handler, etc.)
GTUtility.sendChatTrans(player, "gt.chat.error.machine_broken", machineName);
// lang: gt.chat.error.machine_broken=Your machine %s has broken!
```

> **Singleplayer vs. Multiplayer note:**
> `sendChatTrans` is safe in both environments. In singleplayer the integrated server sends the
> `ChatComponentTranslation` to the local client. On a dedicated server it is sent over the network.
> Either way, translation happens on the client in the player's own language.
> Never pre-translate the string with `StatCollector` before passing it to chat - this breaks

> > multiplayer because translation would happen with the server's (possibly missing) lang file.

Reviewer: Once again: not why that breaks.

It might be clearer to just state that the parameter to `sendChatTrans` must be the translation key, and not a translated value.

Reviewer: ```suggestion
> multiplayer because translation would happen with the server's (possibly wrong) lang file.
```

            .setChatStyle(new ChatStyle().setColor(EnumChatFormatting.GREEN)));
GTUtility.sendChatComp(player, msg);
```

### 1.5 MUI2 - `IKey.lang(key, args...)`
For widgets inside **ModularUI 2** GUIs (`modularui2` package):

```java
import com.cleanroommc.modularui.api.IKey;

// Simple key
IKey.lang("gt.gui.my_label").asWidget()

// Key with placeholders
IKey.lang("gt.gui.tick_rate", tickRate)

// As tooltip line
tooltip.addLine(IKey.lang("gt.gui.my_tooltip"))
```

---

## 2. Text Formatting

Use `EnumChatFormatting` constants. Concatenate them directly with strings.

```java
import net.minecraft.util.EnumChatFormatting;

String text = EnumChatFormatting.RED + "Warning: " + EnumChatFormatting.WHITE + "do not touch";
```

In lang file values you may use `§` codes directly (e.g. `§c` for red, `§r` for reset), but
prefer generating formatting via `EnumChatFormatting` in Java to avoid hard-coded codes.

---

## 3. Format Placeholders

| Placeholder | Use for |
|---|---|
| `%s` | String values (item names, player names, any text) |
| `%d` | Integer / whole number values |
| `%f` or `%.2f` | Floating point values (use sparingly, prefer formatting before passing) |

> | `%1$s`, `%2$s` | Positional placeholders when the order may differ between languages |

Reviewer: This is incorrect, I think, since `ChatComponentTranslation` only understands plain `%s` placeholders.

(Maybe we should patch that with a mixin someday?)

write the remainder in `camelCase`.

File-scoped keys still go under their **normal functional section** in the lang file (GUI keys
under `#### GUI`, info data under `#### Scanner & Info Data`, etc.). The class scope is only
part of the key name - it is not a separate section.

```
gt.<category>.<classScope>.<identifier>
```

| Class | Scope segment |
|---|---|
| `MTEBasicBatteryBuffer` | `basicBatteryBuffer` |
| `MTEMultiBlockBase` | `multiBlockBase` |
| `GTCoverFluidFilter` | `coverFluidFilter` |
| `CoverConveyor` | `coverConveyor` |

**Example** - keys in `MTEBasicBatteryBuffer.java`:

```properties
gt.infodata.basicBatteryBuffer.stored_items=Stored: %s / %s EU
gt.infodata.basicBatteryBuffer.average_input=Avg. Input: %s EU/t
```

If the same key is needed by **more than one file**, drop the class scope segment and use only
the category, making the key shared:

```properties
# shared across multiple machine files
gt.infodata.average_input=Avg. Input: %s EU/t
```

### 4.6 Renaming or Removing Keys
Do not repurpose an existing key for a different string - this silently
breaks other language files that haven't been updated yet. Add a new key

> and remove the old one in the same PR, across all shipped lang files.

Reviewer: This is... unfortunate, to say the least. Not sure we can do better, but I'm not happy with this state of things.

Author: The thing about this, I would say is for the current blank state of the chore. If this would be sorter thru, we could reuse it. (With exception of gui vs chat or server strings)

Reviewer: We might be able to reuse some GUI vs. chat strings eventually. There was that one idea I had about GUI text color substitution via mixin. But this is fine for now.

Author: I wanted to separate the chat vs the gui as of the really one reason and that is coloring. 
But I guess we could always use ColorUtils or or more practically the mui2 themes

---

## 5. Lang File Structure & Section Headers

Comments in `.lang` files start with `#`. Use the following heading depth to organize sections.
This structure mirrors Markdown heading levels and makes the file easier to navigate.

```
#### (4 hashes)  - Top-level / main group
### (3 hashes)   - Sub-group level 1
## (2 hashes)    - Sub-group level 2
# (1 hash)       - Sub-group level 3 (most commonly used for fine-grained sections)

> ```

Reviewer 2: This use an opposite convention to Markdown.

Markdown use `#` for the top level and `##` for sub-group.

Author: This is MarkGH ✨ /j  Big = higher, and yes that just my preferred styling
